sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/util/InternalContext.java

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.generic.core.implementation.util;
+
+import com.generic.core.util.Context;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Internal representation of {@link Context}.
+ */
+public interface InternalContext {
+    /**
+     * Get the key for the internal context.
+     * <p>
+     * For both {@code getKey} and {@code getValue} the value return is the last key-value pair added to the context.
+     *
+     * @return The key for the internal context.
+     */
+    Object getKey();
+
+    /**
+     * Get the value for the internal context.
+     * <p>
+     * For both {@code getKey} and {@code getValue} the value return is the last key-value pair added to the context.
+     *
+     * @return The value for the internal context.
+     */
+    Object getValue();
+
+    /**
+     * Get the number of key-value pairs in the internal context.
+     *
+     * @return The number of key-value pairs in the internal context.
+     */
+    int count();
+
+    /**
+     * Adds a new key-value pair to the internal context.
+     * <p>
+     * Internal context is immutable (shallow only), so this method will return a new instance of the internal context
+     * with the new key-value pair added.
+     *
+     * @param key The key to add.
+     * @param value The value to add.
+     * @return A new instance of the internal context with the new key-value pair added.
+     */
+    InternalContext addData(Object key, Object value);
+
+    /**
+     * Get the value for the given key.
+     * <p>
+     * If the key is not found in the internal context an empty {@link Optional} will be returned.
+     *
+     * @param key The key to get the value for.
+     * @return The value for the given key, or an empty {@link Optional} if the key is not found.
+     */
+    Optional<Object> getData(Object key);
+
+    /**
+     * Get all the key-value pairs in the internal context.
+     *
+     * @param map The map to populate with the key-value pairs.
+     */
+    void getValues(LinkedHashMap<Object, Object> map);
+
+    /**
+     * Merge the given internal context with this internal context.
+     *
+     * @param other The internal context to merge with this internal context.
+     * @return The merged internal context.
+     */
+    InternalContext merge(InternalContext other);
```

Comment:
We don't have a usecase for this yet. We can add this later if needed and it's internal, so, adding it later without `default` is okay.